onnx/defs/math/defs.cc

```diff
@@ -3415,6 +3415,72 @@ static T get_scalar_value_from_tensor(const ONNX_NAMESPACE::TensorProto* t) {
   }
 }
 
+// Shape inference for DFT is shared with IDFT
+void discrete_fourier_transform_shape_inference(ONNX_NAMESPACE::InferenceContext& ctx, bool is_onesided)
+{
+    propagateElemTypeFromInputToOutput(ctx, 0, 0);
+    if (!hasInputShape(ctx, 0))
+    {
+        // If no shape is available for the input, skip shape inference...
+        return;
+    }
+
+    // In general the output shape will match the input shape exactly
+    // So initialize the output shape with the input shape
+    auto& input_shape = getInputShape(ctx, 0);
+    ONNX_NAMESPACE::TensorShapeProto result_shape_proto = input_shape;
+
+    // Get the axis where the DFT will be performed.
+    auto axis = static_cast<int>(getAttribute(ctx, "axis", 1));
+    auto axis_idx = axis % input_shape.dim_size();
+
+    // If dft_length is specified, then we should honor the shape.
+    // Set the output dimension to match the dft_length on the axis.
+    // If onesided this will be adjusted later on...
+    const auto* dft_length = ctx.getInputData(1);
+    if (nullptr != dft_length) {
+        if (dft_length->dims_size() != 0) {
+            fail_shape_inference("dft_length input must be a scalar.");
+        }
+        auto dft_length_value = get_scalar_value_from_tensor<int64_t>(dft_length);
+        result_shape_proto.mutable_dim(axis_idx)->set_dim_value(dft_length_value);
+    }
+
+    // When DFT is onesided, the output shape is half the size of the input shape
+    // along the specified axis.                
+    if (is_onesided) {
+        auto axis_dimension = result_shape_proto.dim(axis_idx);
+        // We need to update the output shape dimension along the specified axis,
+        // but sometimes the dimension will be a free dimension or be otherwise unset.
+        // Only perform inference when a input dimension value exists.
+        if (axis_dimension.has_dim_value())
+        {
+            auto original_signal_size = axis_dimension.dim_value();
+            auto half_signal_size = (original_signal_size >> 1) + 1;
+            result_shape_proto.mutable_dim(axis_idx)->set_dim_value(half_signal_size);
+        }  else
+        {
+            // Clear the value and param (which would otherwie be inherited from the input).
+            result_shape_proto.mutable_dim(axis_idx)->clear_dim_value();
+            result_shape_proto.mutable_dim(axis_idx)->clear_dim_param();
+        }
+    }
+
+    auto dim_size = static_cast<int64_t>(result_shape_proto.dim_size());
+    auto has_component_dimension = dim_size > 2;
+    if (has_component_dimension) {
+        // If the rank is greater than 2, then the last dimension of the input is always the component dimension.
+        // Since the output is always complex, we can just coerce the last dimension to 2.
+        result_shape_proto.mutable_dim(static_cast<int>(dim_size - 1))->set_dim_value(2);
+    } else {
+        // If equal to 2, the last dimensions is missing, and this is the single dimension, real-valued DFT case
```

Comment:
Why this special case handling? Is there some compatibility or convenience reason? Whether the input is real or complex is typically independent of the rank of the input tensor. We could have 2D or 3D real-values, for example. Using different conventions for 2D reals and 3D reals is not a good idea.

Reply:
Uniform convention would be always use one extra dimension (with a value of 1 for real-inputs and 2 for complex-inputs), regardless of whether rank is 2 or greater.

Reply:
This is here because the existing contirib op in onnxruntime allows the special case of 1d reals that do not have a complex dimension. The general rule however is to always have a complex dimension.

Reply:
This will likely cause confusion. I think we should try to make the spec simple. I suggest that we choose between the two simple options we have:

Option 1: Always use 1 extra-dimension as discussed above (no special cases).
Option 2: Always use a (new) attribute to indicate if the input is real or complex. If the input is real, no extra-dimension is used, and an extra-dimension is used only for complex.

Will neither of these options work?

Reply:
Option1 works. I only documented the behavior with the extra dimension.
I can remove the shape inference special case - which is a little hidden easter egg right now.